R/rc_input_event_amendment.R

```diff
@@ -0,0 +1,136 @@
+#' Calculate the monthly timing of carbon inputs for different fertilizer strategies
+#'
+#' This function calculates the timing of carbon inputs (kg C per ha) based on type of organic matter amendments and land use.
+#'
+#' @param crops (data.table) Table with crop rotation, year and potential Carbon inputs.
+#' @param amendment (data.table) A table with the following column names: year, month, cin_tot, cin_hum, cin_dpm, cin_rpm and the fraction eoc over p (fr_eoc_p). Month is optional.
+#'
+#' @details This function increases temporal detail for time series of C inputs of organic amendments.
+#' The inputs for organic amendments are organised in the data.table amendment, where the carbon inputs has the unit kg C / ha.
+#'
+#' The output is an EVENT object.
+#'
+#' @export
+rc_input_event_amendment <- function(crops,amendment = NULL){
+  
+  # add visual bindings
+  B_LU = B_LU_NAME = p_cat = fr_eoc_p = crflt = tcf = NULL
+  cin_hum = cin_rpm = cin_dpm = method = crop_code = crop_name = NULL
+  fr_eoc_p = time = NULL
+  
+  # make local copy
+  dt <- copy(amendment)
+  
+  # make default crop amendment data.table when dt = NULL
+  if(is.null(dt)){dt <- data.table(year = crops[1,year], month = 1, cin_tot = 0, cin_hum = 0,
+                                   cin_dpm = 0, cin_rpm = 0, fr_eoc_p = 10)}
+  
+  # do checks on the crop list
+  checkmate::assert_data_table(crops)
+  checkmate::assert_true('B_LU' %in% colnames(crops))
+  checkmate::assert_true('year' %in% colnames(crops))
+  
+  # do checks on the input of C due to organic amendments
+  checkmate::assert_data_table(dt)
+  checkmate::assert_names(names(dt),
+    must.include = c('year','cin_hum','cin_dpm','cin_rpm','fr_eoc_p'),
+    subset.of    = c('year','month','p_name','cin_tot','cin_hum','cin_dpm','cin_rpm','fr_eoc_p'))
+  checkmate::assert_numeric(dt$cin_hum,lower = 0, upper = 100000,len = nrow(dt))
+  if("cin_tot" %in% names(dt)){
+  checkmate::assert_numeric(dt$cin_tot,lower = 0, upper = 100000,len = nrow(dt))}
+  checkmate::assert_numeric(dt$cin_dpm,lower = 0, upper = 100000,len = nrow(dt))
+  checkmate::assert_numeric(dt$cin_rpm,lower = 0, upper = 100000,len = nrow(dt))
+  checkmate::assert_numeric(dt$fr_eoc_p,lower = 0, upper = 250,len = nrow(dt))
+  checkmate::assert_integerish(dt$year,len = nrow(dt))
+  
+  # load internal tables
+  cc.crop <- as.data.table(rotsee::rc_crops)
+  cc.crop[,B_LU := paste0('nl_',crop_code)]
+  
+  # merge table with amendment c input with crop name
+  dt <- merge(dt, crops[,list(B_LU,year)], by='year')
+  dt <- merge(dt, cc.crop[,list(B_LU,B_LU_NAME=crop_name)],by='B_LU')
```

Comment:
when doing a left join, always add, all.x = TRUE. This ensures that all original rows of dt remain in the merge (otherwise they are automatically removed)